### PR TITLE
Experiment recover

### PR DESCRIPTION
**Pull request by kelunik**
@LeanderK We need a way to delete populations for recovery, everything else should be fine.

### CODE REVIEW

File: src/main/java/edu/kit/ipd/crowdcontrol/objectservice/rest/resources/ExperimentResource.java [hunk 1]
```diff
@@ -126,45 +128,80 @@ public Experiment put(Request request, Response response) {
...
-        experimentOperations.storeRatingOptions(experiment.getRatingOptionsList(), id);
+            try {
+                List<ConstraintRecord> constraints = TagConstraintTransformer.getConstraints(experiment, id);
 
-        populationsHelper.storePopulations(id, experiment.getPopulationsList());
+                constraints.stream()
+                        .filter(constraintRecord -> !constraintRecord.getConstraint().isEmpty())
+                        .forEach(tagConstraintsOperations::insertConstraint);
+            } catch (DataAccessException e) {
+                throw new InternalServerErrorException("Couldn't persist tags.", e);
```
Comment (marcelhollerbach) [inline src/main/java/edu/kit/ipd/crowdcontrol/objectservice/rest/resources/ExperimentResource.java]: can you change this to constraints?

Comment (kelunik) [inline src/main/java/edu/kit/ipd/crowdcontrol/objectservice/rest/resources/ExperimentResource.java]: Done.

**Comment by marcelhollerbach** [comment]
I think a few methods can be refactored to increase readability. Nested try´s arent really nice to read. Do we need to rush here ?

**Comment by kelunik** [comment]
It's a bug, but no critical bug. Will just result in a incomplete experiment, which is actually fine for DRAFT state.

File: src/main/java/edu/kit/ipd/crowdcontrol/objectservice/rest/resources/ExperimentResource.java [hunk 2]
```diff
@@ -318,6 +355,29 @@ private Experiment updateExperimentInfoDraftState(int id, Experiment experiment,
...
+        } catch (DataAccessException e) {
+            // restore old state
+            updateLinkedExperimentInfo(id, old, experiment, oldRecord);
+
+            throw new InternalServerErrorException("Updating the experiment info failed, restored old state.", e);
+        }
+
+        try {
+            //update the experiment itself
+            experimentOperations.updateExperiment(experimentRecord);
```
Comment (marcelhollerbach) [inline src/main/java/edu/kit/ipd/crowdcontrol/objectservice/rest/resources/ExperimentResource.java:369]: Can you move those two try´s into one ? The exact fact if info or experiment failed can be seen in the bt. So we have duplicated code here.

File: src/main/java/edu/kit/ipd/crowdcontrol/objectservice/rest/resources/ExperimentResource.java [hunk 3]
```diff
@@ -126,45 +128,80 @@ public Experiment put(Request request, Response response) {
...
+        } catch (DataAccessException e) {
+            throw new InternalServerErrorException("Experiment could not be stored.", e);
+        }
 
-        tags.stream()
-                .filter(tagRecord -> !tagRecord.getTag().isEmpty())
-                .forEach(tagConstraintsOperations::insertTag);
+        try {
+            try {
+                List<TagRecord> tags = TagConstraintTransformer.getTags(experiment, id);
```
Comment (marcelhollerbach) [inline src/main/java/edu/kit/ipd/crowdcontrol/objectservice/rest/resources/ExperimentResource.java:141]: I am quite not sure if this should be its own function which creates the info of a expeirment, makes reading easier

**Comment by kelunik** [comment]
Still misses platform recovery by @LeanderK.

**Comment by LeanderK** [comment]
what do you need?

**Comment by kelunik** [comment]
We need to restore platforms from before the change. Currently we just add platforms in `storePopulations`. But in case something failed, I don't think we want it to appear in the database at all.

**Comment by LeanderK** [comment]
well this is more complicated then one would think. I'll look into it when i have time

**Comment by kelunik** [comment]
Yes, that's why it's not done yet. We could make it the last thing we change.

**Comment by LeanderK** [comment]
there are still some bugs left ;)

**Comment by kelunik** [comment]
Hm?

**Comment by LeanderK** [comment]
the multiple-calibrations bug, simon mentioned that he received an invalid reservation error at the demo and when submitting only one rating was not possible. I'll create some tickets

**Comment by LeanderK** [comment]
i may try to do this tomorrow

**Comment by LeanderK** [comment]
obviously this feature was not fully realised, but the progress must not be thrown away. I would create a issue with a short description and a reference to this pull request, so that future work can start here.

**Pull request closed without merging**